sdk/search/Azure.Search/src/Serialization/JsonExtensions.cs

```diff
@@ -0,0 +1,473 @@
+        {
+            // Short circuit for empty/erroneous streams.
+            if (json == null)
+            {
+                return default;
+            }
+
+            int written = 0;
+            byte[] rented = null;
+            ReadOnlySpan<byte> utf8Bom = Constants.Utf8Bom;
+            try
+            {
+                if (json.CanSeek)
+                {
+                    // Ask for 1 more than the length to avoid resizing later,
+                    // which is unnecessary in the common case where the stream
+                    // length doesn't change.
+                    long expectedLength = Math.Max(utf8Bom.Length, json.Length - json.Position) + 1;
+                    rented = ArrayPool<byte>.Shared.Rent(checked((int)expectedLength));
+                }
+                else
+                {
+                    rented = ArrayPool<byte>.Shared.Rent(Constants.UnseekableStreamInitialRentSize);
+                }
+
+                // Read up to 3 bytes to see if it's the UTF-8 BOM
+                int lastRead;
+                do
+                {
+                    // No need for checking for growth, the minimal rent sizes
+                    // both guarantee it'll fit.
+                    Debug.Assert(rented.Length >= utf8Bom.Length);
+                    lastRead = json.Read(rented, written, utf8Bom.Length - written);
+                    written += lastRead;
+                } while (lastRead > 0 && written < utf8Bom.Length);
+
+                // If we have 3 bytes, and they're the BOM, reset the write
+                // position to 0.
+                if (written == utf8Bom.Length &&
+                    utf8Bom.SequenceEqual(rented.AsSpan(0, utf8Bom.Length)))
+                {
+                    written = 0;
+                }
+
+                do
+                {
+                    if (rented.Length == written)
+                    {
+                        byte[] toReturn = rented;
+                        rented = ArrayPool<byte>.Shared.Rent(checked(toReturn.Length * 2));
+                        Buffer.BlockCopy(toReturn, 0, rented, 0, toReturn.Length);
+
+                        // Holds document content, clear it.
+                        ArrayPool<byte>.Shared.Return(toReturn, clearArray: true);
+                    }
+
+                    lastRead = json.Read(rented, written, rented.Length - written);
+                    written += lastRead;
+                } while (lastRead > 0);
+
+                // Deserialize the JSON once we've copied it over
+                return JsonSerializer.Deserialize<T>(rented.AsSpan(0, written), JsonExtensions.SerializerOptions);
+            }
+            finally
+            {
+                if (rented != null)
+                {
+                    // Holds document content, clear it before returning it.
+                    rented.AsSpan(0, written).Clear();
+                    ArrayPool<byte>.Shared.Return(rented);
```

Comment:
I think Return has bool clear parameter.

Reply:
Yes - I thought the same thing when I saw that at first.  It uses Return to clear earlier in the code, but here it only clears however much it's used in case you only used a couple of bytes in a giant buffer.

Reply:
That would clear all of `rented` which might be up to 2x larger than needed (in the worst case).